README
This is a collection of the unix tools that nobody thought to write
long ago, when unix was young. Currently it consists of these tools:

chronic: runs a command quietly unless it fails
combine: combine the lines in two files using boolean operations
ifdata: get network interface info without parsing ifconfig output
isutf8: check if a file or standard input is utf-8
ifne: run a command if the standard input is not empty
lckdo: execute a program with a lock held (deprecated)
mispipe: pipe two commands, returning the exit status of the first
parallel: run multiple jobs at once
pee: tee standard input to pipes
sponge: soak up standard input and write to a file
ts: timestamp standard input
vidir: edit a directory in your text editor
vipe: insert a text editor into a pipe
zrun: automatically uncompress arguments to command

Its web page is here: http://kitenet.net/~joey/code/moreutils/

Your suggestions of additional tools to add to this collection are
appreciated. The web page lists some that are under consideration but
have not yet been included, I also welcome feedback on which of these to
include.

-- Joey Hess <joey@kitenet.net>


debian/changelog
moreutils (0.45) unstable; urgency=low

  * ts: Support %.s for seconds sinch epoch with subsecond resolution.
    Closes: #619764

 -- Joey Hess <joeyh@debian.org>  Sun, 19 Jun 2011 15:44:28 -0400

moreutils (0.44) unstable; urgency=low

  * pee: Propigate exit status of commands run.

 -- Joey Hess <joeyh@debian.org>  Thu, 10 Mar 2011 17:38:14 -0400

moreutils (0.43) unstable; urgency=low

  * chronic: New command, runs a command quietly, unless it fails.
  * Now depends on IPC::Run, used by chronic.

 -- Joey Hess <joeyh@debian.org>  Fri, 29 Oct 2010 15:54:37 -0400

moreutils (0.42) unstable; urgency=low

  * sponge: Guarantee that output file is always updated atomically,
    by renaming a temp file into place. Closes: #592144
  * sponge: Ensure that output file permissions are always preserved
    if it already exists.
  * Typo. Closes: #596032

 -- Joey Hess <joeyh@debian.org>  Wed, 06 Oct 2010 20:03:25 -0400

moreutils (0.41) unstable; urgency=low

  * ifdata.docbook: Mark interface as required in synopsis. Closes: #588397
  * Add missing AUTHOR section to docbook man pages.
  * sponge: Correct bad use of fread that caused a trailing quantity of
    soaked data to be silently discarded when a temp file was used
    and sponge output to stdout. Closes: #595220

 -- Joey Hess <joeyh@debian.org>  Thu, 02 Sep 2010 16:47:10 -0400

moreutils (0.40) unstable; urgency=low

  * lckdo: Now deprecated, since util-linux's flock(1) can do the same
    thing.
  * parallel: -i will now replace {} inside parameters, before the {} had
    to be a separate parameter.

 -- Joey Hess <joeyh@debian.org>  Tue, 06 Jul 2010 19:38:14 -0400

moreutils (0.39) unstable; urgency=low

  * parallel: Fix exit code handling when commands are specified after --
  * parallel: Make -j 0 do something reasonable (start all jobs at once).
  * parallel: Fix to really avoid running new jobs when load is too high.
  * parallel: Fix logic error in code handling -l that could make parallel
    return a bogus 255 exit code when all jobs succeeded. Closes: #569617
  * parallel: Allow a decimal load value to be specified with -l
  * Caps sillyness. Closes: #570815
  * zrun: Add support for .xz files.

 -- Joey Hess <joeyh@debian.org>  Tue, 23 Feb 2010 15:51:26 -0500

moreutils (0.38) unstable; urgency=low

  * Description improvements. Closes: #549450
    (Thanks, Justin B Rye)
  * parallel: Allow running independent commands,
    like `parallel -j3 -- ls df "echo hi"`
  * ifdata: Add FreeBSD kernel support, although some of the more esoteric
    interface options are not currently supported in FreeBSD.
  * parallel: Define WEXITED to allow building on FreeBSD kernel.
  * Thanks Enrico Tassi for the FreeBSD kernel support, which should
    be enough to get moreutils built on Debian kFreeBSD. Closes: #562609

 -- Joey Hess <joeyh@debian.org>  Tue, 09 Feb 2010 15:50:42 -0500

moreutils (0.37) unstable; urgency=low

  * parallel: Clarify man page regarding CPUs. Closes: #536597
  * parallel: Add -n option. Thanks, Pierre Habouzit. Closes: #537992
    (As a side effect, fixes a segfault if -- was omitted.)
  * parallel.1: Typo fixes. Closes: #538147

 -- Joey Hess <joeyh@debian.org>  Sat, 25 Jul 2009 10:18:40 +0200

moreutils (0.36) unstable; urgency=low

  * parallel: New program, contributed by Tollef Fog Heen,
    that can run multiple jobs in parallel, optionally checking
    load average.
  * mispipe: Fix closing of extra pipe FD before starting command
    so it is not inherited by daemons. Closes: #533448
    (Thanks, Jeremie Koenig)

 -- Joey Hess <joeyh@debian.org>  Fri, 10 Jul 2009 11:01:41 -0400

moreutils (0.35) unstable; urgency=low

  * ifdata: Don't assume that all interface names are 6 characters or less,
    for instance "wmaster0" is longer. Increase the limit to 20 characters.
    Closes: #526654 (Thanks, Alan Pope)
  * isutf8: Reject UTF-8-encoded UTF-16 surrogates. Closes: #525301
    (Thanks, Jakub Wilk and liw)

 -- Joey Hess <joeyh@debian.org>  Tue, 05 May 2009 15:06:41 -0400

moreutils (0.34) unstable; urgency=low

  * vipe: Avoid dying on empty input. Thanks, Anders Kaseorg
    Closes: #508491

 -- Joey Hess <joeyh@debian.org>  Thu, 11 Dec 2008 15:11:23 -0500

moreutils (0.33) unstable; urgency=low

  * Support installing moreutils into prefixes other than /usr (Evan Broder)
  * Fix zrun breakage introduced last version. Closes: #504129

 -- Joey Hess <joeyh@debian.org>  Fri, 31 Oct 2008 17:01:03 -0400

moreutils (0.32) unstable; urgency=low

  * zrun: Can be linked to zsomeprog to run the equivilant of zrun someprog.
    Closes: #411623 (Stefan Fritsch)
  * zrun: Add support for lzma and lzo. (Stefan Fritsch)
  * Fix pod error in vidir(1).

 -- Joey Hess <joeyh@debian.org>  Sun, 26 Oct 2008 23:43:30 -0400

moreutils (0.31) unstable; urgency=low

  * pee.1: Document difference with tee in stdout.
  * ts: Support displaying fractional seconds via a "%.S" conversion
    specification. Closes: #482789

 -- Joey Hess <joeyh@debian.org>  Sat, 28 Jun 2008 23:19:31 -0400

moreutils (0.30) unstable; urgency=low

  * debhelper v7; rules file minimisation
  * Use DESTDIR instead of PREFIX.
  * Add a DOCBOOK2XMAN setting. (Greg KH)
  * ifne: Add -n which makes it run the command if stdin is empty.
  * ifne: If no command is specified, print usage information.

 -- Joey Hess <joeyh@debian.org>  Wed, 14 May 2008 19:37:42 -0400

moreutils (0.29) unstable; urgency=low

  * Add ifne, contributed by Javier Merino.
  * sponge, ifne: Ensure that suspending/resuming doesn't
    result in partial writes of the data, by using fwrite()
    rather than write().
  * sponge: Handle large data sizes by using a temp file rather than by
    consuming arbitrary amounts of memory. Patch by Brock Noland.
  * ts: Allow both -r and a format to be specified, to parse dates and output
    in a specified format.
  * ts: Fix bug in timezone regexp.

 -- Joey Hess <joeyh@debian.org>  Tue, 15 Apr 2008 15:30:52 -0400


isutf8.c
/*
 * isutf8.c - do the input files look like valid utf-8 byte streams?
 * 
 * Copyright (C) 2005  Lars Wirzenius
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
 */

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <getopt.h>


#define VERSION "1.1"


/*
 * Code to indicate an invalid UTF8 character.
 */
enum { INVALID_CHAR = 0xffffffff };


/*
 * Produce shortest UTF8 encoding of a 31-bit value in 'u', returning it
 * in the array 'buf'. Return the number of bytes in the encoded value.
 * If the value is too large (more than 32 bits or would take more than
 * 'maxbytes' bytes), return -1.
 */
static int encodeutf8(unsigned long u, unsigned char *buf, size_t maxbytes)
{
        static const struct {
            int nbytes;
            unsigned long max;
        } tab[] = {
            { 1, 0x0000007F },
            { 2, 0x000007FF },
            { 3, 0x0000FFFF },
            { 4, 0x001FFFFF },
            { 5, 0x03FFFFFF },
            { 6, 0x7FFFFFFF },
        };
        static const int ntab = sizeof(tab) / sizeof(tab[0]);
        int i, j;

        if (u > tab[ntab-1].max)
                return -1;

        for (i = 0; i < ntab; ++i) {
                if (u <= tab[i].max)
                    break;
        }
        assert(i < ntab);

        if (tab[i].nbytes > maxbytes)
                return -1;
        
        if (tab[i].nbytes == 1) { /* Special case */
                buf[0] = u;
        } else {
                for (j = tab[i].nbytes-1; j > 0; --j) {
                        buf[j] = 0x80 | (u & 0x3f);
                        u >>= 6;
                }
        
                unsigned char mask = ~(0xFF >> tab[i].nbytes);
                buf[0] = mask | u;
        }

        return tab[i].nbytes;
}


/* 
 * Return number of ones at the top of a byte.
 *
 * I'm pretty sure there is a fancy trick to do this without a loop,
 * but I'm too tired to figure it out now. --liw
 */
static int high_ones(int c) {
        int n;

        for (n = 0; (c & 0x80) == 0x80; c <<= 1)
                ++n;    
        return n;
}


/*
 * Decode a UTF8 character from an array of bytes. Return character code.
 * Upon error, return INVALID_CHAR.
 */
static unsigned long decodeutf8(unsigned char *buf, int nbytes)
{
        unsigned long u;
        int i, j;
        
        if (nbytes <= 0)
                return INVALID_CHAR;
        
        if (nbytes == 1) {
                if (buf[0] >= 0x80)
                        return INVALID_CHAR;
                return buf[0];
        }
        
        i = high_ones(buf[0]);
        if (i != nbytes)
                return INVALID_CHAR;    
        u = buf[0] & (0xff >> i);
        for (j = 1; j < nbytes; ++j) {
                if ((buf[j] & 0xC0) != 0x80)
                            return INVALID_CHAR;
                u = (u << 6) | (buf[j] & 0x3f);
        }

        /* Conforming UTF-8 cannot contain codes 0xd800–0xdfff (UTF-16 
           surrogates) as well as 0xfffe and 0xffff. */
        if (u >= 0xD800 && u <= 0xDFFF)
            return INVALID_CHAR;
        if (u == 0xFFFE || u == 0xFFFF)
            return INVALID_CHAR;

        return u;
}


/*
 * Determine if the contents of an open file form a valid UTF8 byte stream.
 * Do this by collecting bytes for a character into a buffer and then
 * decode the bytes and re-encode them and compare that they are identical
 * to the original bytes. If any step fails, return 0 for error. If EOF
 * is reached, return 1 for OK.
 */
static int is_utf8_byte_stream(FILE *file, char *filename, int quiet) {
        enum { MAX_UTF8_BYTES = 6 };
        unsigned char buf[MAX_UTF8_BYTES];
        unsigned char buf2[MAX_UTF8_BYTES];
        int nbytes, nbytes2;
        int c;
        unsigned long code;
        unsigned long line, col, byteoff;

        nbytes = 0;
        line = 1;
        col = 1;
        byteoff = 0;
                
        for (;;) {
                c = getc(file);
    
                if (c == EOF || c < 0x80 || (c & 0xC0) != 0x80) {
                        /* New char starts, deal with previous one. */
                        if (nbytes > 0) {
                                code = decodeutf8(buf, nbytes);
                                if (code == INVALID_CHAR)
                                        goto error;
                                nbytes2 = encodeutf8(code, buf2, 
                                                     MAX_UTF8_BYTES);
                                if (nbytes != nbytes2 || 
                                    memcmp(buf, buf2, nbytes) != 0)
                                        goto error;
                                ++col;
                        }
                        nbytes = 0;
                        /* If it's UTF8, start collecting again. */
                        if (c != EOF && c >= 0x80)
                                buf[nbytes++] = c;
                } else {
                        /* This is a continuation byte, append to buffer. */
                        if (nbytes == MAX_UTF8_BYTES)
                                goto error;
                        buf[nbytes++] = c;
                }
    
                if (c == EOF)
                        break;
                else if (c == '\n') {
                        ++line;
                        byteoff = 0;
                        col = 1;
                } else
                        ++byteoff;
        }
        
        if (nbytes != 0)
                goto error;

	return 1;
	
error:
	if (!quiet) {
		printf("%s: line %lu, char %lu, byte offset %lu: "
		       "invalid UTF-8 code\n", filename, line, col, byteoff);
	}
	return 0;
}


static void usage(const char *program_name) {
	printf("Usage: %s [-hq] [--help] [--quiet] [file ...]\n", 
	       program_name);
	printf("Check whether input files are valid UTF-8.\n");
	printf("This is version %s.\n", VERSION);
}


int main(int argc, char **argv) {
	int i, ok;
	FILE *file;

	int quiet;
	struct option options[] = {
		{ "help", no_argument, NULL, 'h' },
		{ "quiet", no_argument, &quiet, 1 },
	};
	int opt;
	
	quiet = 0;
	
	while ((opt = getopt_long(argc, argv, "hq", options, NULL)) != -1) {
		switch (opt) {
		case 0:
			break;
			
		case 'h':
			usage(argv[0]);
			exit(0);
			break;
			
		case 'q':
			quiet = 1;
			break;

		case '?':
			exit(EXIT_FAILURE);

		default:
			abort();
		}
	}

	if (optind == argc)
		ok = is_utf8_byte_stream(stdin, "stdin", quiet);
	else {
		ok = 1;
		for (i = optind; i < argc; ++i) {
			file = fopen(argv[i], "r");
			if (file == NULL) {
				fprintf(stderr, "isutf8: %s: error %d: %s\n", 
				                argv[i], errno, 
				                strerror(errno));
				ok = 0;
			} else {
			        if (! is_utf8_byte_stream(file, argv[i], quiet))
			            ok = 0;
				(void) fclose(file);
			}
		}
	}
	
	if (ok)
		exit(0);
	exit(EXIT_FAILURE);
}
